// ares-webos-sdk/lib/cli-appdata.js
/*jshint node: true, strict: false, globalstrict: false */

var
    async = require('async'),
    fs = require('fs'),
    log = require('npmlog'),
    mkdirp = require('mkdirp'),
    path = require('path'),
    shelljs = require('shelljs'),
    Validator = require('jsonschema').Validator;

var
    clone = require('./util/objclone').deepCopy;

var Cli = (function() {
    var instance;

    function finish(err, returnValue, next) {
        if (next && typeof next === 'function') {
            if (!returnValue) {
                return setImmediate(next, err);
            }
            return setImmediate(next, err, returnValue);
        }
        return err || returnValue;
    }

    function init() {
        var
            workDir,
            builtinDeviceListFile,
            userDeviceListFile,
            workParentDir = path.resolve(process.env.APPDATA || process.env.HOME || process.env.USERPROFILE),
            builtinFiles = {
            "config" : {
                "dir" : path.join(__dirname, "json"),
                "file" : "config.json"
            },
            "commands" : {
                "dir" : path.join(__dirname, "json"),
                "file" : "command-service.json"
            },
            "deviceList" : {
                "dir" : path.join(__dirname, "json"),
                "file" : "novacom-devices.json"
            },
            "deviceListSchema" : {
                "dir" : path.join(__dirname, "schema"),
                "file" : "NovacomDevices.schema"
            }
        };

        for (key in builtinFiles) {
            var filePath = path.join( builtinFiles[key].dir, builtinFiles[key].file );
            (function(store, key, file) {
                store[key] = JSON.parse(fs.readFileSync(file));
                store[key + 'File'] = filePath;
            })(this, key, filePath);
        }

        workDir = (this.config.dataDir)? path.join(workParentDir, this.config.dataDir) : path.join(workParentDir, '.webos');
        builtinDeviceListFile = path.join(builtinFiles["deviceList"].dir, builtinFiles["deviceList"].file);
        userDeviceListFile = path.join(workDir, builtinFiles["deviceList"].file);

        if (!fs.existsSync(workDir)) {
            mkdirp.sync(workDir);
        }
        try {
            var userDeviceList = JSON.parse(fs.readFileSync(userDeviceListFile));
            var result = validateDevice(this.deviceListSchema, userDeviceList);
            if (result && result.resultValue) {
                this['deviceList'] = userDeviceList;
            } else {
                log.verbose("cliAppData() schema check failure:", userDeviceListFile);
                log.verbose("cliAppData() result:", result);
                log.verbose("cliAppData() copy " + builtinDeviceListFile + " to " + userDeviceListFile);
                shelljs.cp('-f', builtinDeviceListFile, userDeviceListFile);
            }
        } catch (err) {
            log.verbose("cliAppData()#err:", err);
            log.verbose("cliAppData() copy " + builtinDeviceListFile + " to " + userDeviceListFile);
            shelljs.cp('-f', builtinDeviceListFile, userDeviceListFile);
        }

        this['deviceListFile'] = userDeviceListFile;
        this['workDir'] = workDir;
        return true;
    }

    function validateDevice(schema, data) {
        var result = {
            resultValue: false,
            err: null
        };
        var schemaArray = {
            "id": "test",
            "type": "array",
            "items": {
                "$ref": "/deviceSchema"
            }
        };
        var v = new Validator();
        try {
            v.addSchema(schema, "/deviceSchema");
            var result =  v.validate(data, schemaArray);
            if (result && result.errors.length > 0) {
                var errMsg = "Invalid device info."
                for (idx in result.errors) {
                    errMsg = errMsg.concat("\n");
                    var errMsgLine = result.errors[idx].property + " "
                        + result.errors[idx].message;
                    var regex = /instance\[*.*\]*\./g;
                    if ((result = regex.exec(errMsgLine)) != null) {
                        errMsgLine = errMsgLine.substring(result[0].length);
                    }
                    errMsg = errMsg.concat(errMsgLine);
                }
                result.err = new Error(errMsg);
            } else {
                result.resultValue = true;
            }
        } catch (err) {
            //ignore exception
            result.err = err;
        }
        return result;
    }

    function CliAppData() {
        if (instance) {
            return instance;
        }
        instance = this;
        if (!this.inialized) {
            this.inialized = init.call(this);
        }
    }

    CliAppData.prototype = {
        getPath: function(next) {
            return finish(null, this.workDir, next);
        },

        getAppDir: function(next) {
            return finish(null, path.join(__dirname, '..'), next);
        },

        getConfig: function(next) {
            return finish(null, clone(this.config), next);
        },

        compareProfile: function(query, next) {
            next(null, query.toLowerCase() === this.config.profile.toLowerCase());
        },

        compareProfileSync: function(query){
            return query.toLowerCase() === this.config.profile.toLowerCase();
        },

        getDeviceList: function(renewal, next) {
            return finish(null, clone(this.deviceList), next);
        },

        getCommandService: function (next) {
            return finish(null, clone(this.commands), next);
        },

        setDeviceList: function(deviceListData, next) {
            try{
                var result = validateDevice(this.deviceListSchema, deviceListData);
                if (result && !result.resultValue) {
                    var err = result.err || "invalid data";
                    return finish(err, this.deviceList, next);
                }
                this.deviceList = deviceListData;
                fs.writeFileSync(this.deviceListFile, JSON.stringify(this.deviceList, null, '\t'));
            } catch (err) {
                return finish(err, this.deviceList, next);
            }
            return finish(null, this.deviceList, next);
        },

        setConfig: function (configData, next) {
            try{
                this.config = configData;
                fs.writeFileSync(this.configFile, JSON.stringify(this.config, null, '\t'));
            } catch (err) {
                return finish(err, this.config, next);
            }
            return finish(null, this.config, next);
        }
    }
    return CliAppData;
}());

if (typeof module !== 'undefined' && module.exports) {
    module.exports = Cli;
}
